File: old_csharp_ver/PowerBlock/HeartBeat.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net;
using System.IO;
using System.Threading;

namespace PowerBlock
{
    class HeartBeat
    {
        private static bool ShownOnce = false;
        private string Request = "http://www.minecraft.net/heartbeat.jsp?port=PORT&max=MAX&name=SERVER-NAME&public=True&version=7&salt=SALT&users=USERS";

        public HeartBeat()
        {
            Request = Request.Replace("PORT", CraftServer.Port.ToString());
            Request = Request.Replace("MAX", CraftServer.MaxPlayers.ToString());
            Request = Request.Replace("SERVER-NAME", CraftServer.Name);
            Request = Request.Replace("SALT", CraftServer.Salt);
            Request = Request.Replace("USERS", CraftServer.Players.Length.ToString());
            CraftEvent.CraftHeartbeatEvent(Request);
            ThreadPool.QueueUserWorkItem(new WaitCallback(AsyncRequest));
        }

        public void AsyncRequest(object o)
        {
            try
            {
                WebClient wc = new WebClient();
                string Response = wc.DownloadString(Request);
                if (Response.StartsWith("http://www.minecraft.net") == false)
                {
                    Console.WriteLine("Failed to send heartbeat! Is minecraft.net down?");
                }
                else
                {
                    FileStream fs = new FileStream(CraftServer.Environment + CraftServer.DirStr + "external-url.txt", FileMode.Create);
                    StreamWriter sw = new StreamWriter(fs);
                    sw.WriteLine(Response);
                    sw.Flush();
                    sw.Close();
                    fs.Close();
                    if (ShownOnce == false)
                    {
                        Console.WriteLine("Saved HeartBeat URL to external-url.txt!");
                        ShownOnce = true;
                    }
                }
            }
            catch (Exception HeartBeatException)
            {
                Console.WriteLine("Failed to send heartbeat! Is minecraft.net down?\n\t" + HeartBeatException.Message);
            }
        }
    }
}

File: com/craftera/server/Heartbeat.java
package com.craftera.server;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.InputStreamReader;
import java.net.URL;

public class Heartbeat {

	private int port;
	private int max;
	private String name;
	private String isPublic;
	private String salt;
	private int users;
	private URL url;
	
	public Heartbeat() {
		try {
			StringBuilder sb = new StringBuilder();
			sb.append("http://www.minecraft.net/heartbeat.jsp?");
			sb.append("name=" + name + "&");
			sb.append("port=" + String.valueOf(port) + "&");
			sb.append("max=" + String.valueOf(max) + "&");
			sb.append("public=" + isPublic + "&");
			sb.append("version=7&");
			sb.append("salt=" + salt + "&");
			sb.append("users=" + String.valueOf(users));
			url = new URL(sb.toString());
		}
		catch (Exception ex) { }
	}
	
	public String send() {
		String response = null;
		try {
			response = new BufferedReader(new InputStreamReader(url.openStream())).readLine();
			if (!(response.indexOf("bad heartbeat") >= 0))
				System.out.println("Heartbeat sent! " + response + "URL also saved to /game-url.txt");
			else
				throw new Exception();
		}
		catch (Exception ex) { System.out.println("Could not send heartbeat! Is minecraft.net unavailable?"); }
		try {
			BufferedWriter bw = new BufferedWriter(new FileWriter(System.getProperty("user.dir") + File.separator + "game-url.txt"));
			bw.write(response + "\n");
			bw.flush();
			bw.close();
		}
		catch (Exception ex) { }
		return response;
	}
}
